cdap/resource_local_artifact.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cdap
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func resourceLocalArtifact() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceLocalArtifactCreate,
+		Read:   resourceArtifactRead,
+		Delete: resourceArtifactDelete,
+		Exists: resourceArtifactExists,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"namespace": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				DefaultFunc: func() (interface{}, error) {
+					return defaultNamespace, nil
+				},
+			},
+			// Technically, we could omit the version in the API call because CDAP will infer the
+			// version from the jar. However, forcing the user to specify the version makes dealing
+			// with the resource easier because other API calls require it.
+			"version": {
```

Comment:
Computing it ourselves means reading the jar's manifest file, like the server will do:

https://docs.cask.co/cdap/4.0.0/en/reference-manual/http-restful-api/artifact.html

I think this is possible, and probably better, but it might be something that we can improve in a follow up (or as needs arise)?

Reply:
Yeah, I definitely think it's something we should do given the main purpose of this resource was not manually having to set fields that are already set in the config file.

It doesn't need to be a blockerfor now, but it is something that should be done soon as it is a breaking API change. If you don't get to it in this PR I can take it on.

Reply:
You actually have to unpack the jar and read the manifest in order to get this info (see link). It's doable, but more involved than reading the config file.

Reply:
Ah, got it. I thought it was in the json config for some reason. Ok, can be manual for now.